### PR TITLE
WIP: Cache operators

### PR DESCRIPTION
**Pull request by nchristensen**
Needs some debugging to pass test_tri_diff_mat.

### CODE REVIEW

**Comment by nchristensen** [comment]
test_tri_diff_mat output:

```
INFO     test_grudge:test_grudge.py:200 axis 0
h    | Error                  | Running EOC      
-----+------------------------+-------------------
0.1  | 3.3231299588987184e-05 |                  
0.05 | 1.674878784996281e-06  | 4.310414121000102

Overall EOC: 4.310414121000102
INFO     test_grudge:test_grudge.py:200 axis 1
h    | Error             | Running EOC           
-----+-------------------+------------------------
0.1  | 5.999996501107085 |                       
0.05 | 5.999999823747682 | -7.989264296381368e-07

Overall EOC: -7.989264296381368e-07
```

I'm not certain yet why this is happening. The norm of the difference between the cached and the generated matrix is equal to zero at each iteration.

File: grudge/execution.py [hunk 1]
```diff
@@ -530,14 +530,33 @@ def prg(nmatrices):
...
+                matrices_ary_dev = self.bound_op.operator_data_cache[cache_key]
+            except KeyError:
+                matrices = repr_op.matrices(out_grp, in_grp)
+                matrices_ary = np.empty((
+                    noperators, out_grp.nunit_dofs, in_grp.nunit_dofs))
+                for i, op in enumerate(insn.operators):
+                    matrices_ary[i] = matrices[op.rst_axis]
+                matrices_ary_dev = self.array_context.from_numpy(matrices_ary)
+                self.bound_op.operator_data_cache[cache_key] = matrices_ary_dev
+            cached = matrices_ary_dev.get()
```
Comment (inducer) [inline grudge/execution.py:546]: You could start by asserting that the thing being pulled out of the cache is always the same as what would be computed "fresh".

Comment (nchristensen) [reply grudge/execution.py:546]: I do something similar below but I print the norm of the difference. It is always zero.

Comment (nchristensen) [reply grudge/execution.py:546]: Looks like there are a few non-zeros after all.

```
0.0
0.0
0.0
0.0
0.0
0.0
0.8164965809277261
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.8164965809277261
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.8164965809277261
0.0
0.0
0.0
0.0
0.0
```

Comment (nchristensen) [reply grudge/execution.py:546]: Somehow the key must not be unique. It fails on the key `('diff_batch', <meshmode.discretization.poly_element.PolynomialWarpAndBlendElementGroup object at 0x7f3fd5afea30>, <meshmode.discretization.poly_element.PolynomialWarpAndBlendElementGroup object at 0x7f3fd5afea30>, RefDiffOperator(1, DOFDesc(DTAG_VOLUME_ALL, QTAG_NONE), DOFDesc(DTAG_VOLUME_ALL, QTAG_NONE)), dtype('O'))`.

The expected matrix is a re-ordering of the cached matrix. 

```
DIFFERENCE
[[[ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00
    0.00000000e+00  0.00000000e+00]
  [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00
    0.00000000e+00  0.00000000e+00]
  [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00
    0.00000000e+00  0.00000000e+00]
  ...
  [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00
    0.00000000e+00  0.00000000e+00]
  [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00
    0.00000000e+00  0.00000000e+00]
  [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00
    0.00000000e+00  0.00000000e+00]]

 [[ 1.77635684e-15 -6.75650249e+00  2.66666667e+00 ...  5.40952719e-17
   -4.60668587e-17 -5.00000000e-01]
  [ 2.55900472e-01 -3.33484533e+00 -2.12826024e+00 ...  4.70287863e-01
   -5.00643726e-16 -6.90212516e-01]
  [ 1.38947513e-01 -1.53875832e-01 -3.19693019e+00 ...  1.25557455e+00
    7.68075300e-16 -8.32871903e-01]
  ...
  [ 2.22044605e-16  1.77635684e-15  9.43689571e-16 ... -2.66453526e-15
   -5.56298892e-16  1.24099025e+00]
  [ 5.55111512e-16 -6.47568470e-16 -4.02514616e-16 ... -2.41013806e+00
    3.33484533e+00  9.85089781e-01]
  [-5.55111512e-17 -5.24311481e-16  1.08528733e-15 ... -6.75650249e+00
    1.62143398e-15  5.00000000e+00]]

 [[-1.77635684e-15  6.75650249e+00 -2.66666667e+00 ... -5.40952719e-17
    4.60668587e-17  5.00000000e-01]
  [-2.55900472e-01  3.33484533e+00  2.12826024e+00 ... -4.70287863e-01
    5.00643726e-16  6.90212516e-01]
  [-1.38947513e-01  1.53875832e-01  3.19693019e+00 ... -1.25557455e+00
   -7.68075300e-16  8.32871903e-01]
  ...
  [-2.22044605e-16 -1.77635684e-15 -9.43689571e-16 ...  2.66453526e-15
    5.56298892e-16 -1.24099025e+00]
  [-5.55111512e-16  6.47568470e-16  4.02514616e-16 ...  2.41013806e+00
   -3.33484533e+00 -9.85089781e-01]
  [ 5.55111512e-17  5.24311481e-16 -1.08528733e-15 ...  6.75650249e+00
   -1.62143398e-15 -5.00000000e+00]]]
CACHED
[[[-5.00000000e+00  1.08535204e-15  1.23740375e-15 ...  1.83045093e-16
    4.68282467e-16 -8.87005072e-17]
  [-9.85089781e-01 -3.33484533e+00 -3.82517118e-01 ... -1.38873957e-16
    2.78011437e-16  1.05075241e-16]
  [ 5.13947513e-01 -1.49046041e+00 -3.19693019e+00 ... -2.48435269e-16
   -2.48435269e-16  1.49116705e-16]
  ...
  [ 6.90212516e-01 -4.70287863e-01 -1.71966137e-01 ...  1.18622941e-15
    2.41013806e+00 -3.14247705e-16]
  [ 6.90212516e-01 -1.77204815e-16 -5.23391918e-17 ... -1.62839388e-16
    3.33484533e+00 -2.55900472e-01]
  [ 5.00000000e-01  1.07333163e-15 -1.49727373e-16 ... -1.22611980e-15
    6.75650249e+00  6.02260136e-16]]

 [[-5.00000000e+00 -1.16585736e-15  1.53960760e-15 ...  1.41494222e-16
    5.23575428e-17 -5.00000000e-01]
  [-9.85089781e-01 -3.33484533e+00 -3.82517118e-01 ...  4.70287863e-01
   -4.71217885e-16 -6.90212516e-01]
  [ 5.13947513e-01 -1.49046041e+00 -3.19693019e+00 ...  1.25557455e+00
    7.85363142e-16 -8.32871903e-01]
  ...
  [ 6.90212516e-01 -4.70287863e-01 -1.71966137e-01 ...  3.33484533e+00
   -6.54469285e-16  9.85089781e-01]
  [ 6.90212516e-01  1.50956526e-16 -7.73921543e-16 ... -4.62972782e-16
    3.33484533e+00  9.85089781e-01]
  [ 5.00000000e-01  1.20657858e-15 -1.14172806e-15 ...  1.15851285e-16
    1.88487154e-15  5.00000000e+00]]

 [[-5.00000000e+00  6.75650249e+00 -2.66666667e+00 ...  8.73989506e-17
    9.84244016e-17 -8.98907502e-17]
  [-1.24099025e+00 -2.78588958e-15  1.74574312e+00 ...  8.08490162e-17
    2.94258403e-17 -3.16587862e-17]
  [ 3.75000000e-01 -1.33658458e+00  1.01777395e-16 ... -9.43953045e-18
    1.72878427e-17 -1.69789549e-17]
  ...
  [ 6.90212516e-01 -4.70287863e-01 -1.71966137e-01 ...  3.33484533e+00
   -9.81703928e-17 -2.55900472e-01]
  [ 6.90212516e-01  7.98524996e-16 -3.71406928e-16 ...  2.41013806e+00
    1.55417241e-16 -5.15628173e-17]
  [ 5.00000000e-01  1.73089006e-15 -2.22701539e-15 ...  6.75650249e+00
    2.63437558e-16  4.43222777e-16]]]
UNCACHED
[[[-5.00000000e+00  1.08535204e-15  1.23740375e-15 ...  1.83045093e-16
    4.68282467e-16 -8.87005072e-17]
  [-9.85089781e-01 -3.33484533e+00 -3.82517118e-01 ... -1.38873957e-16
    2.78011437e-16  1.05075241e-16]
  [ 5.13947513e-01 -1.49046041e+00 -3.19693019e+00 ... -2.48435269e-16
   -2.48435269e-16  1.49116705e-16]
  ...
  [ 6.90212516e-01 -4.70287863e-01 -1.71966137e-01 ...  1.18622941e-15
    2.41013806e+00 -3.14247705e-16]
  [ 6.90212516e-01 -1.77204815e-16 -5.23391918e-17 ... -1.62839388e-16
    3.33484533e+00 -2.55900472e-01]
  [ 5.00000000e-01  1.07333163e-15 -1.49727373e-16 ... -1.22611980e-15
    6.75650249e+00  6.02260136e-16]]

 [[-5.00000000e+00  6.75650249e+00 -2.66666667e+00 ...  8.73989506e-17
    9.84244016e-17 -8.98907502e-17]
  [-1.24099025e+00 -2.78588958e-15  1.74574312e+00 ...  8.08490162e-17
    2.94258403e-17 -3.16587862e-17]
  [ 3.75000000e-01 -1.33658458e+00  1.01777395e-16 ... -9.43953045e-18
    1.72878427e-17 -1.69789549e-17]
  ...
  [ 6.90212516e-01 -4.70287863e-01 -1.71966137e-01 ...  3.33484533e+00
   -9.81703928e-17 -2.55900472e-01]
  [ 6.90212516e-01  7.98524996e-16 -3.71406928e-16 ...  2.41013806e+00
    1.55417241e-16 -5.15628173e-17]
  [ 5.00000000e-01  1.73089006e-15 -2.22701539e-15 ...  6.75650249e+00
    2.63437558e-16  4.43222777e-16]]

 [[-5.00000000e+00 -1.16585736e-15  1.53960760e-15 ...  1.41494222e-16
    5.23575428e-17 -5.00000000e-01]
  [-9.85089781e-01 -3.33484533e+00 -3.82517118e-01 ...  4.70287863e-01
   -4.71217885e-16 -6.90212516e-01]
  [ 5.13947513e-01 -1.49046041e+00 -3.19693019e+00 ...  1.25557455e+00
    7.85363142e-16 -8.32871903e-01]
  ...
  [ 6.90212516e-01 -4.70287863e-01 -1.71966137e-01 ...  3.33484533e+00
   -6.54469285e-16  9.85089781e-01]
  [ 6.90212516e-01  1.50956526e-16 -7.73921543e-16 ... -4.62972782e-16
    3.33484533e+00  9.85089781e-01]
  [ 5.00000000e-01  1.20657858e-15 -1.14172806e-15 ...  1.15851285e-16
    1.88487154e-15  5.00000000e+00]]]

```

File: grudge/execution.py [hunk 2]
```diff
@@ -530,14 +530,33 @@ def prg(nmatrices):
             if in_grp.nelements == 0:
                 continue
 
+            # Cache operator
+            cache_key = ("diff_batch", in_grp, out_grp, repr_op, field.dtype)
```
Comment (inducer) [inline grudge/execution.py:534]: You need to consider all operators in the cache key, not just the first one.

Comment (nchristensen) [reply grudge/execution.py:534]: Yeah, that seems to do it.

**Comment by nchristensen** [comment]
Closing this pull request one in favor of #19

**Pull request closed without merging**